src/components/tl/ucp/scatter/scatter_knomial.c

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "tl_ucp_coll.h"
+#include "tl_ucp_sendrecv.h"
+#include "core/ucc_progress_queue.h"
+#include "core/ucc_mc.h"
+#include "coll_patterns/sra_knomial.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+
+#define SAVE_STATE(_phase)                                                    \
+    do {                                                                      \
+        task->scatter_kn.phase = _phase;                                      \
+    } while (0)
+
+#define GET_BASE_PEER(_radix, _rank, _dist, _peer)                            \
+    do {                                                                      \
+        _peer = _rank - ((_rank / _dist) % _radix) * _dist;                   \
+} while (0)
+
+enum {
+    UCC_SCATTER_KN_PHASE_INIT,
+    UCC_SCATTER_KN_PHASE_LOOP, /* main loop of recursive k-ing */
+};
+
+ucc_rank_t calc_recv_dist(ucc_rank_t team_size, ucc_rank_t rank,
+                                     ucc_rank_t radix, ucc_rank_t root)
+{
+	if (rank == root) {
+		return 0;
+	}
+    ucc_rank_t root_base;
+    ucc_rank_t dist = 1;
+    GET_BASE_PEER(radix, root, dist, root_base);
+    while (dist <= team_size) {
+        if (rank >= root_base && rank < root_base + radix * dist) {
+            break;
+        }
+        dist *= radix;
+        GET_BASE_PEER(radix, root_base, dist, root_base);
+    }
+    return dist;
+}
+
+ucc_status_t
```

Comment:
plz provide brief description of the algorithm: compute dist, single recv, several sends for non-leafs, etc